src/main/java/com/microsoft/azure/spring/data/cosmosdb/core/convert/MappingDocumentDbConverter.java

```diff
@@ -47,6 +48,47 @@ public MappingDocumentDbConverter(
         this.objectMapper = objectMapper == null ? ObjectMapperFactory.getObjectMapper() : objectMapper;
     }
 
+    /**
+     * Convert a property value to the value stored in CosmosDB
+     *
+     * @param fromPropertyValue
+     * @return
+     */
+    public static Object toDocumentDBValue(Object fromPropertyValue) {
+        if (fromPropertyValue == null) {
+            return null;
+        }
+
+        // com.microsoft.azure.documentdb.JsonSerializable#set(String, T) cannot set values for Date and Enum correctly
+        if (fromPropertyValue instanceof Date) {
+            fromPropertyValue = ((Date) fromPropertyValue).getTime();
+        } else if (fromPropertyValue instanceof Enum) {
+            fromPropertyValue = fromPropertyValue.toString();
+        }
+
+        return fromPropertyValue;
+    }
+
+    public <T extends Object> T readAsync(Class<T> domainClass,
+                                          @NonNull com.microsoft.azure.cosmosdb.Document document) {
+        objectMapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
+
+        final DocumentDbPersistentEntity<?> entity = mappingContext.getPersistentEntity(domainClass);
+        Assert.notNull(mappingContext.getPersistentEntity(domainClass), "entity not be null");
+
+        final DocumentDbPersistentProperty idProperty = entity.getIdProperty();
+        Assert.notNull(idProperty, "idProperty should not be null");
+
+        final JSONObject json = new JSONObject(document.toJson()).put(idProperty.getName(), document.getId());
+
+        try {
+            return objectMapper.readValue(json.toString(), domainClass);
+        } catch (IOException e) {
+            throw new IllegalStateException("Failed to read the source document " + document.toJson()
```

Comment:
Why not use `DocumentDBAccessException` here but IllegalState directly?

Reply:
fixed.